### PR TITLE
debounce rounding position on Marker

### LINKED ISSUE #11135: Add isHighAccurate property to Marker
## Motivation
For animating a point the example that Mapbox provided is [animate point along a route](https://docs.mapbox.com/mapbox-gl-js/example/animate-point-along-route/). This convention has bad performance and triggers a repaint map on every update source.  
image
as you can see The CPU is involved throughout the airplane movement.

If we use Marker instead of symbol layer the result is very different, because it works with CSS and doesn't trigger a repaint. 
image
As you can see this convention uses less CPU and memory, but if you check this you can see [Codesandbox link](https://codesandbox.io/s/beautiful-dew-ummh0?file=/index.html) that movement is not smooth and sometimes the marker shakes. This behavior happens because round `marker._pos` property of Marker(maybe for performance) but if we want to implement animation movement, we should have highAccuracy.

### Implementation
We can add `isHighAccuracy` property and check if we want high accurate marker don't round `marker._pos`.

### PR DESCRIPTION
**Pull request by malekeym**
## Launch Checklist
In this PR we debounce rounding position for cases that we need to animate `Marker`
Close #11135

 - [x] briefly describe the changes in this PR
 - [ ] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Animated markers are smoother in most browsers (h/t [@malekeym](https://github.com/malekeym))</changelog>`

### CODE REVIEW

File: src/ui/marker.js
Comment (SnailBones) [reply src/ui/marker.js:544]: It's true that we don't really want postponed snapping on `move` events, but I'm not sure if that would actually cause an issue. All `move` events should be followed by either another `move` or `moveend` event, so I think the delayed rounding should be canceled by your call to  `window.cancelAnimationFrame`. There might be cases (maybe due to latency) where the event isn't called in time, so this needs testing. (If it proves to be an issue, I think we could solve it by adding a third option for `move` events to do no rounding at all)

The other way sounds similar to our discussion on https://github.com/mapbox/mapbox-gl-js/pull/11136. I think it's a viable idea and I like its simplicity in the code, but as @mourner mentioned we prefer to avoid adding extra user options unless there is no alternative. Ideally, we want this to work without developers even having to consider it.

> My main reason for proposing this option is to avoid expanding the API surface for something that can be an internal detail users don't have to think about. If we can make it to _just work_ without any user-side changes for both cases, that's ideal.

My preference would be to continue this current PR, and keep the extra option idea as a backup if we're sure that this doesn't meet your needs. What do you think? Is there something I'm missing about why the other approach might better meet your use case?

With either approach, I think we could simplify the code by changing `_update()` to take in a boolean instead of `e?: {type: 'move' | 'moveend'}`

Comment (malekeym) [reply src/ui/marker.js:544]: Thank you @SnailBones, Yes it's clear.
Ok for this PR should I change anything?

Comment (SnailBones) [reply src/ui/marker.js:544]: Thank you @malekeym!
Yes, do the changes I mention [above](https://github.com/mapbox/mapbox-gl-js/pull/11167#discussion_r736046490) make sense to you? Let me know if you need more clarification.
This will also need some unit tests before we can merge it in.

Comment (malekeym) [reply src/ui/marker.js]: @SnailBones Yes, it's clear, I think changing `_update` to accept `boolean` is better, I change the code, I'll appreciate to check changes and let me know if something is missed or wrong, if all things are ok, let me know to add unit tests. Thank you.

Comment (SnailBones) [reply src/ui/marker.js]: Great work! This looks good to me and I think it's ready for unit tests.

Comment (malekeym) [reply src/ui/marker.js]: Thank you @SnailBones I add unit tests, too.

Comment (SnailBones) [reply src/ui/marker.js]: This is a great start @malekeym! Can you also add tests to ensure that the snap is _not_ happening immediately on `setLngLat` on map move?

Comment (malekeym) [reply src/ui/marker.js]: @SnailBones I add some unit tests for `move` and `moveend` and `setLngLat`, could you check, and if another cases that my test not cover let me know.

**Review by SnailBones — CHANGES_REQUESTED** [review]
The unit tests look great to me!

I just tested this out in browser, and unfortunately the bug still isn't fixed.
https://user-images.githubusercontent.com/14878684/141167735-a2739059-b954-409e-9082-e4cd513fb15a.mov

~~I think this is because of the `map._forceMarkerUpdate()`, which I added to fix another marker bug [here](https://github.com/mapbox/mapbox-gl-js/pull/10985). The intention was to have markers update in response to `map.setTerrain()`, but it seems to be calling excess updates as well. I think that fixing this will require checking if terrain actually changes on the call to `painter.updateTerrain()` and only updating the markers if so.~~
EDIT: I'm wrong, this isn't the cause of the bug.

**Comment by malekeym** [comment]
> The unit tests look great to me!
> 
> I just tested this out in browser, and unfortunately the bug still isn't fixed. https://user-images.githubusercontent.com/14878684/141167735-a2739059-b954-409e-9082-e4cd513fb15a.mov
> 
> I think this is because of the `map._forceMarkerUpdate()`, which I added to fix another marker bug [here](https://github.com/mapbox/mapbox-gl-js/pull/10985). The intention was to have markers update in response to `map.setTerrain()`, but it seems to be calling excess updates as well. I think that fixing this will require checking if terrain actually changes on the call to `painter.updateTerrain()` and only updating the markers if so.


https://user-images.githubusercontent.com/70053533/141185587-57a65470-6e5b-444b-90dd-c9e22be11fe2.mov

@SnailBones good point, I think I catch it. but when i test it in my browser all thing is fine. 
could you provide a code that demonstrates this bug.

**Comment by SnailBones** [comment]
> @SnailBones good point, I think I catch it. but when i test it in my browser all thing is fine. could you provide a code that demonstrates this bug.

I've copied the code from [this example](https://docs.mapbox.com/mapbox-gl-js/example/animate-marker/) into a local html file, and tested in both Firefox and Chrome.

Your video unfortunately doesn't seem to be working for me.

**Comment by malekeym** [comment]
> > @SnailBones good point, I think I catch it. but when i test it in my browser all thing is fine. could you provide a code that demonstrates this bug.
> 
> I've copied the code from [this example](https://docs.mapbox.com/mapbox-gl-js/example/animate-marker/) into a local html file, and tested in both Firefox and Chrome.
> 
> Your video unfortunately doesn't seem to be working for me.

actually i copied that code...
https://user-images.githubusercontent.com/70053533/141190959-299cad85-3bcc-4622-9787-6ef5c25e95e9.mov
Could you check this link?

**Comment by SnailBones** [comment]
> actually i copied that code... https://user-images.githubusercontent.com/70053533/141190959-299cad85-3bcc-4622-9787-6ef5c25e95e9.mov Could you check this link?

Still not working for me. I see this:
image

**Comment by mourner** [comment]
@SnailBones here's a better format:

https://user-images.githubusercontent.com/25395/141192449-cbc4ee67-6ddb-4a09-9c8a-04da569ada69.mp4

File: src/ui/marker.js
Comment (mourner) [inline src/ui/marker.js:286]: This won't actually remove the listener that was added previously, because this is a different function. You need to pass the exact reference — e.g. by creating a separate `_updateMoving` that internally calls `_update(true)`.

**Comment by SnailBones** [comment]
Thanks for the conversion @mourner! 
@malekeym It's hard for me to tell if the marker is snapping in that video. I've found that the bug is most visible when a marker is moving very slowly, like this:

https://user-images.githubusercontent.com/14878684/141192992-4a778959-14a3-4dc2-a3dc-bff8cfed8f56.mov

Here's the code I used to create this:

```js
<!DOCTYPE html>
<html>
<head>
    <title>Marker snapping test</title>
    <meta charset='utf-8'>
    <link rel='stylesheet' href='../dist/mapbox-gl.css' />
    <style>
        body { margin: 0; padding: 0; }
        html, body, #map { height: 100%; }
    </style>
</head>

<body>
<div id='map'></div>
<script src='../dist/mapbox-gl-dev.js'></script>
<script src='../debug/access_token_generated.js'></script>
<script>

var map = new mapboxgl.Map({
    container: 'map',
    style: 'mapbox://styles/mapbox/streets-v11',
    zoom: 0
});

const marker = new mapboxgl.Marker({color: '#F84C4C'});

function animateMarker(timestamp) {
    const radius = 5;
    marker.setLngLat([
        Math.cos(timestamp / 1000) * radius,
        Math.sin(timestamp / 1000) * radius
    ]);
    marker.addTo(map);
    requestAnimationFrame(animateMarker);
}

requestAnimationFrame(animateMarker);

</script>
</body>
</html>

```

**Review by mourner — CHANGES_REQUESTED** [review]

**Comment by malekeym** [comment]
> Thanks for the conversion @mourner! @malekeym It's hard for me to tell if the marker is snapping in that video. I've found that the bug is most visible when a marker is moving very slowly, like this:
> 
>  Screen.Recording.2021-11-10.at.1.59.58.PM.mov 
> Here's the code I used to create this:
> 
> ```js
> <!DOCTYPE html>
> <html>
> <head>
>     <title>Marker snapping test</title>
>     <meta charset='utf-8'>
>     <link rel='stylesheet' href='../dist/mapbox-gl.css' />
>     <style>
>         body { margin: 0; padding: 0; }
>         html, body, #map { height: 100%; }
>     </style>
> </head>
> 
> <body>
> <div id='map'></div>
> <script src='../dist/mapbox-gl-dev.js'></script>
> <script src='../debug/access_token_generated.js'></script>
> <script>
> 
> var map = new mapboxgl.Map({
>     container: 'map',
>     style: 'mapbox://styles/mapbox/streets-v11',
>     zoom: 0
> });
> 
> const marker = new mapboxgl.Marker({color: '#F84C4C'});
> 
> function animateMarker(timestamp) {
>     const radius = 5;
>     marker.setLngLat([
>         Math.cos(timestamp / 1000) * radius,
>         Math.sin(timestamp / 1000) * radius
>     ]);
>     marker.addTo(map);
>     requestAnimationFrame(animateMarker);
> }
> 
> requestAnimationFrame(animateMarker);
> 
> </script>
> </body>
> </html>
> ```
Thank you @SnailBones , I check the code.  if you see in function `animateMarker`  call `addTo` after `setLngLat` so the result is that _pos is rounded, as we discussed before we decided to round position in `addTo` and `moveend`. am i right?

File: src/ui/marker.js
Comment (malekeym) [reply src/ui/marker.js]: Done!

**Comment by SnailBones** [comment]
> Thank you @SnailBones , I check the code. if you see in function `animateMarker` call `addTo` after `setLngLat` so the result is that _pos is rounded, as we discussed before we decided to round position immediately in `addTo` and `moveend`. am i right?

Good catch @malekeym! That works for me, you're correct that the issue results from the repeated calls to `marker.addTo`. One last suggestion before I approve this: as a way to fix this specific case, can we add an early exit condition to the `addTo()`? Maybe something like:
```js
        if (map === this._map) { return this; }
```

**Comment by malekeym** [comment]
> > Thank you @SnailBones , I check the code. if you see in function `animateMarker` call `addTo` after `setLngLat` so the result is that _pos is rounded, as we discussed before we decided to round position immediately in `addTo` and `moveend`. am i right?
> 
> Good catch @malekeym! That works for me, you're correct that the issue results from the repeated calls to `marker.addTo`. One last suggestion before I approve this: as a way to fix this specific case, can we add an early exit condition to the `addTo()`? Maybe something like:
> 
> ```js
>         if (map === this._map) { return this; }
> ```

@SnailBones very good idea, i changed the code, but i find a bug that is related to `_forceUpdateMarker` :))  if you want to reproduce change the code that i sent and load terrain Map instead of street view, then move the map you can see the marker is jagging. (this is not happening in street view), could you help me to fix this?

**Comment by SnailBones** [comment]
> I found the bug that is related to @mourner [Review](https://github.com/mapbox/mapbox-gl-js/pull/11167#discussion_r746984287), I don't know why but when using `_updateMoving` the behaviour is changed on the `move` event. To demonstrate the issue first create a map then just add a Marker and move the map. Is it clear?

I'm glad you noticed that issue @malekeym! The cause is that `this` is bound to the map in `updateMoving`, so it's actually calling `map._update` If you can find a way to correctly call `marker._update` the bug should be fixed.

**Comment by malekeym** [comment]
> > I found the bug that is related to @mourner [Review](https://github.com/mapbox/mapbox-gl-js/pull/11167#discussion_r746984287), I don't know why but when using `_updateMoving` the behaviour is changed on the `move` event. To demonstrate the issue first create a map then just add a Marker and move the map. Is it clear?
> 
> I'm glad you noticed that issue @malekeym! The cause is that `this` is bound to the map in `updateMoving`, so it's actually calling `map._update` If you can find a way to correctly call `marker._update` the bug should be fixed.

Thank you @SnailBones for finding the reason that bug happened I made changes and add a unit test to check move events and the issue is fixed but I have a problem with flow test, I can fix it with define property `_updateMove` instead of method and use Arrow function, or I should use wild card for disabling flow on reassigning Method. I think the first solution is better what do you think?

**Comment by SnailBones** [comment]
I think you have the right idea with assigning a property with an arrow function! Because besides the flow error, I think the current approach runs into the same issue that @mourner mentioned above because `.bind();` creates a new method.

**Comment by malekeym** [comment]
> I think you have the right idea with assigning a property with an arrow function! Because besides the flow error, I think the current approach runs into the same issue that @mourner mentioned above because `.bind();` creates a new method.

I use the arrow function, but just one question, if we use bind and then assign the new function to a method, so the marker references are the same, so the @mourner  issue is fixed, am I right?

**Review by SnailBones — APPROVED** [review]
Looks good to me!

**Comment by malekeym** [comment]
@SnailBones any blocker for this merge request? should I change anything?

**Review by mourner — APPROVED** [review]
This looks great to me now!

**Comment by malekeym** [comment]
@SnailBones and @mourner , when I merge the main branch to my branch, I see that `DOM` is removed and used js for add style so I changed my code, too.

File: src/ui/marker.js
Comment (malekeym) [reply src/ui/marker.js]: Yes, but I don't have any idea, do you have?

Comment (SnailBones) [reply src/ui/marker.js]: Maybe seperate that line into something like:
```js
function updateTransform(){   DOM.setTransform(this._element, `${anchorTranslate[this._anchor]} translate(${this._pos.x}px, ${this._pos.y}px) ${pitch} ${rotation}`);}
```

Comment (malekeym) [reply src/ui/marker.js]: I separate this line as `_updateTransform`, Do you think it's a good idea to move pitch and rotation calculation into that method?

File: src/ui/marker.js
Comment (SnailBones) [inline src/ui/marker.js:455]: Great work @malekeym!
One more tiny nit: Why does this function take in the position as an argument? Given it's directly accessing class variables like `this._anchor`, I think that it would be more consistent and prevent potential confusion to either remove this argument, or else pass in the other variables as arguments.

Comment (malekeym) [reply src/ui/marker.js]: @SnailBones Done!

Comment (SnailBones) [reply src/ui/marker.js]: Looks great!

**Comment by SnailBones** [comment]
Merging it in, thanks for your work on this @malekeym!  :pray:

**Pull request merged**
